lib/evmone_precompiles/ecc.hpp

```diff
@@ -0,0 +1,229 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2023 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#include <evmmax/evmmax.hpp>
+
+namespace evmmax::ecc
+{
+
+/// The affine (two coordinates) point on an Elliptic Curve over a prime field.
+template <typename IntT>
+struct Point
+{
+    IntT x = 0;
+    IntT y = 0;
+
+    friend constexpr bool operator==(const Point& a, const Point& b) noexcept
+    {
+        // TODO(intx): C++20 operator<=> = default does not work for uint256.
+        return a.x == b.x && a.y == b.y;
+    }
+
+    /// Checks if the point represents the special "infinity" value.
+    [[nodiscard]] constexpr bool is_inf() const noexcept { return *this == Point{}; }
+};
+
+static_assert(Point<unsigned>{}.is_inf());
+
+template <typename IntT>
+struct ProjPoint
+{
+    IntT x = 0;
+    IntT y = 1;
+    IntT z = 0;
+
+    /// Checks if the point represents the special "infinity" value.
+    [[nodiscard]] constexpr bool is_inf() const noexcept { return x == 0 && z == 0; }
+};
+
+static_assert(ProjPoint<unsigned>{}.is_inf());
+
+template <typename IntT>
+using InvFn = IntT (*)(const ModArith<IntT>&, const IntT& x) noexcept;
+
+/// Converts an affine point to a projected point with coordinates in Montgomery form.
+template <typename IntT>
+inline ProjPoint<IntT> to_proj(const ModArith<IntT>& s, const Point<IntT>& p) noexcept
+{
+    // FIXME: Add to_mont(1) to ModArith?
+    // FIXME: Handle inf
+    return {s.to_mont(p.x), s.to_mont(p.y), s.to_mont(1)};
+}
+
+/// Converts a projected point to an affine point.
+template <typename IntT>
+inline Point<IntT> to_affine(
+    const ModArith<IntT>& s, InvFn<IntT> inv, const ProjPoint<IntT>& p) noexcept
+{
+    // FIXME: Split to_affine() and to/from_mont(). This is not good idea.
+    // FIXME: Add tests for inf.
+    const auto z_inv = inv(s, p.z);
+    return {s.from_mont(s.mul(p.x, z_inv)), s.from_mont(s.mul(p.y, z_inv))};
+}
+
+template <typename IntT, int A = 0>
+ProjPoint<IntT> add(const evmmax::ModArith<IntT>& s, const ProjPoint<IntT>& p,
+    const ProjPoint<IntT>& q, const IntT& b3) noexcept
+{
+    static_assert(A == 0, "point addition procedure is simplified for a = 0");
+
+    if (p.is_inf())
+        return q;
```

Comment:
Shouldn't we return `infinity` here and for other cases when input is at infinity?

Reply:
The "infinity" work as identity element (0) :shrug:

Reply:
Make sense then 👍